app/assets/javascripts/discourse/app/app.js

```diff
@@ -67,15 +70,82 @@ const Discourse = Application.extend({
+        // Ember:
+        //
+        // | Ember                 | Discourse          |                        |
+        // | initializers          | pre-initializers   | runs once per app load |
+        // | instance-initializers | (api-)initializers | runs once per app boot |
+        //
+        // However, addons can also add them to the app in the build, which would
+        // be under the standard Ember naming conventions, so in practice, we will
+        // have a mix of the two, which resulted in an unfortunate naming conflict
+        // we'll have to deal with below.
+
+        if (key.includes("/pre-initializers/")) {
+          const initializer = this._prepareInitializer(key, true);
+          if (initializer) {
+            this.initializer(initializer);
+          }
+        } else if (key.includes("/api-initializers/")) {
+          const initializer = this._prepareInitializer(key, true);
+          if (initializer) {
+            this.instanceInitializer(initializer);
+          }
+        } else if (key.includes("/instance-initializers/")) {
+          const initializer = this._prepareInitializer(key, false);
+          if (initializer) {
+            this.instanceInitializer({
+              name: key.slice(key.lastIndexOf("/") + 1),
+              ...initializer,
+            });
+          }
+        } else if (key.includes("/initializers/")) {
+          // Due to the naming conflict, this can be either.
+          // Assume it isn't discourse style for the time being, to avoid
+          // mutating the module.
+          let initializer = this._prepareInitializer(key, false);
+          if (initializer) {
+            // Coincidentally, Ember default blueprints for `initializers` and
+            // `instance-initializers` do not include a `name` property, but
+            // `this.initializer(...)` and `this.instanceInitializer(...)`
+            // expects them to have one.
+            //
+            // Typically, this is resolved by `ember-load-initializers` which
+            // [infers a name][1] based on the module name.
+            //
+            // On the other hand, Discourse "native" initializers all seem to
+            // have a name explicitly set, probably inheriting the convention
+            // from the "concatenating all the files" days.
+            //
+            // Using this quirk, we can sort of tell if an initializer is the
+            // "Discourse-style" or "Ember-style" by checking for the `name`
+            // property. If there is no name, it's likely generated from the
+            // Ember blueprint and introduced by an addon.
+            //
+            // Obviously, this is very fragile as there is nothing requiring
+            // addons to use the generators or stopping addon authors from
+            // adding a `name` property explicitly. If we infer the wrong
+            // thing here it'll probably cause more subtle issues somewhere
+            // else.
```

Comment:
Yeah I agree this seems pretty fragile. I'd much prefer a check like

```
if(key.startsWith("discourse/"))
```

All Discourse core/theme/plugin initializers are guaranteed to meet that check, so it should be much more robust.